clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -780,79 +782,20 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
-                    cur_instr_info.decoding.is_prefetch =
-                        instr_is_prefetch(noalloc_instr);
-                    cur_instr_info.decoding.opcode = instr_get_opcode(noalloc_instr);
-#ifdef X86
-                    cur_instr_info.decoding.is_xsave = instr_is_xsave(noalloc_instr);
-                    cur_instr_info.decoding.is_xrstor = instr_is_xrstor(noalloc_instr);
-#endif
-                    cur_instr_info.decoding.is_syscall = instr_is_syscall(noalloc_instr);
-                    cur_instr_info.decoding.writes_memory =
-                        instr_writes_memory(noalloc_instr);
-                    cur_instr_info.decoding.is_predicated =
-                        instr_is_predicated(noalloc_instr);
-                    // DR_ISA_REGDEPS instructions don't have an opcode, hence we use
-                    // their category to determine whether they perform at least one read
-                    // or write.
-                    if (instr_get_isa_mode(noalloc_instr) == DR_ISA_REGDEPS) {
-                        cur_instr_info.decoding.num_memory_read_access =
-                            TESTANY(DR_INSTR_CATEGORY_LOAD,
-                                    instr_get_category(noalloc_instr))
-                            ? 1
-                            : 0;
-                        cur_instr_info.decoding.num_memory_write_access =
-                            TESTANY(DR_INSTR_CATEGORY_STORE,
-                                    instr_get_category(noalloc_instr))
-                            ? 1
-                            : 0;
-                    } else {
-                        cur_instr_info.decoding.num_memory_read_access =
-                            instr_num_memory_read_access(noalloc_instr);
-                        cur_instr_info.decoding.num_memory_write_access =
-                            instr_num_memory_write_access(noalloc_instr);
-                    }
-                    if (type_is_instr_branch(memref.instr.type) &&
-                        // DR_ISA_REGDEPS instructions don't have a branch target saved as
-                        // instr.src[0], so we cannot retrieve this information.
-                        !TESTANY(OFFLINE_FILE_TYPE_ARCH_REGDEPS, shard->file_type_)) {
-                        const opnd_t target = instr_get_target(noalloc_instr);
-                        if (opnd_is_pc(target)) {
-                            cur_instr_info.decoding.branch_target =
-                                reinterpret_cast<addr_t>(opnd_get_pc(target));
-                        }
-                    }
-                }
-                shard->decode_cache_[trace_pc] = cur_instr_info.decoding;
-            }
+            shard->decode_cache_.add_decode_info(memref.instr);
+            cur_instr_info.decoding = shard->decode_cache_.get_decode_info(trace_pc);
 #ifdef X86
-            if (cur_instr_info.decoding.opcode == OP_sti)
+            if (cur_instr_info.decoding->opcode == OP_sti)
```

Comment:
Now that nullptr is possible from this interface, this is unsafe and has to check for null.

Reply:
Not a ptr anymore.